src/rgw/rgw_rest_conn.cc

```diff
@@ -69,11 +82,47 @@ int RGWRESTConn::get_url(string& endpoint)
 {
   if (endpoints.empty()) {
     ldout(cct, 0) << "ERROR: endpoints not configured for upstream zone" << dendl;
-    return -EIO;
+    return -EINVAL;
   }
 
-  int i = ++counter;
-  endpoint = endpoints[i % endpoints.size()];
+  size_t num = 0;
+  while (num < endpoints.size()) {
+    int i = ++counter;
+    endpoint = endpoints[i % endpoints.size()];
+
+    if (endpoints_status.find(endpoint) == endpoints_status.end()) {
+      ldout(cct, 1) << "ERROR: missing status for endpoint " << endpoint << dendl;
+      num++;
+      continue;
+    }
+
+    const auto& upd_time = endpoints_status[endpoint].load();
+
+    if (ceph::real_clock::is_zero(upd_time)) {
+      break;
+    }
+
+    auto diff = ceph::to_seconds<double>(ceph::real_clock::now() - upd_time);
+
+    ldout(cct, 20) << "endpoint url=" << endpoint
+                   << " last endpoint status update time="
+                   << ceph::real_clock::to_double(upd_time)
+                   << " diff=" << diff << dendl;
+
+    static constexpr uint32_t CONN_STATUS_EXPIRE_SECS = 2;
+    if (diff >= CONN_STATUS_EXPIRE_SECS) {
```

Comment:
@jzhu116-bloomberg .. could you please explain what it means if connection_status is not expired here (i.e, diff < 2sec). Shouldn't it be considered valid endpoint?

This check added here breaks cloud-transition code and perhaps cloud-sync module too - https://tracker.ceph.com/issues/65251

Reply:
@soumyakoduri this tracks endpoints that recently returned a connection error so we can retry against other endpoints. 'not expired' here means that it failed recently enough that we shouldn't try again

however, this logic is not very good at differentiating between connection errors and other http errors. it just checks for EIO, which turns out to be the default case in [rgw_http_error_to_errno()](https://github.com/ceph/ceph/blob/a2cd8d5/src/rgw/rgw_http_errors.h#L18-L40)

@yuvalif and i discussed this recently in https://github.com/ceph/ceph/pull/55527, and changed some pubsub ops to return ERR_SERVICE_UNAVAILABLE instead of EIO to work around that

i would like to find some other default error code for `rgw_http_error_to_errno()`, but it's not clear what that might break in multisite

for your cloud transition testing, can you tell which earlier request is getting mapped to EIO?

Reply:
> for your cloud transition testing, can you tell which earlier request is getting mapped to EIO?

Thanks Casey. 
For simple object transition case, I can think of requests such as target bucket creation (https://github.com/ceph/ceph/blob/main/src/rgw/driver/rados/rgw_lc_tier.cc#L1217) and fetching HEAD object (https://github.com/ceph/ceph/blob/main/src/rgw/driver/rados/rgw_lc_tier.cc#L1282) which may fail. This is expected at times and hence the errors are ignored.  The errors returned by the cloud endpoint for these ops may include EIO. 

There may be few more such cases with multipart upload to the cloud.

Reply:
> i would like to find some other default error code for rgw_http_error_to_errno(), but it's not clear what that might break in multisite

if we can't prevent these cloud transition ops from returning errors that map to EIO, removing EIO from rgw_http_error_to_errno() seems like the only viable solution. i pushed https://github.com/ceph/ceph/pull/56704 as an example. @smanjara do you think we could reasonably validate such a change with multisite testing at this point?

Reply:
there are places where we return EIO explicitly but I am not sure where we map some other error to EIO. I'll pull your changes from #56704 and run it through our suite @cbodley

Reply:
@smanjara i've added that to a test batch already. i'll share the results and we can go through them together

Reply:
Thanks Casey and Shilpa